cdc/metadata.cc

```diff
@@ -86,33 +93,31 @@ cdc::stream_id cdc::metadata::get_stream(api::timestamp_type ts, dht::token tok)
         // Nothing protects us from that until we start using transactions for generation switching.
     }
 
-    auto it = gen_used_at(now);
+    auto it = gen_used_at(ts);
     if (it == _gens.end()) {
         throw std::runtime_error(format(
-                "cdc::metadata::get_stream: could not find any CDC stream (current time: {})."
-                " Are we in the middle of a cluster upgrade?", format_timestamp(now)));
+                "cdc::metadata::get_stream: could not find any CDC stream ({})."
+                " Are we in the middle of a cluster upgrade?", format_timestamp(ts)));
     }
 
-    // Garbage-collect generations that will no longer be used.
-    it = _gens.erase(_gens.begin(), it);
+    auto current_gen = it == gen_used_at(now);
+
+    // Garbage-collect generations that will no longer be used. If `it` points to a generation older than
+    // the one operating at `now - generation_leeway`, it will be invalidated.
+    clean_gens(now - generation_leeway.count());
```

Comment:
There are 3 calls to `upper_bound` now, which is a relatively complex algorithm. This is hot path code, called on every write to CDC-log-enabled table.

Would it be possible to rewrite this logic to have only one `upper_bound` call?

For example, start with `it = gen_used_at(ts)`, then iterate backwards (using second iterator variable) until we hit the beginning or generation at `now - generation_leeway` and erase gens up to that point.

Eventually after a generation change (and relatively quickly, after `generation_leeway` passes) we will clean all gens other than the "current" one so this iteration logic (triggered on every write) will finish at `it == begin()` and thus will be very cheap.

Reply:
Sounds good, I'll try changing the code this way.

Reply:
> For example, start with `it = gen_used_at(ts)`, then iterate backwards (using second iterator variable) until we hit the beginning or generation at `now - generation_leeway` and erase gens up to that point.

If `gen_used_at(ts)` is older than  `gen_used_at(now - generation_leeway)`, we won't remove generations `[gen_used_at(ts), gen_used_at(now - generation_leeway))` even though we should. Fixing this requires also iterating forward in some cases which complicates the code. However, this should be a very rare case and future writes will remove these obsolete generations anyway, so maybe it's not a problem.

If we disallow all writes behind `now - generation_leeway`, there will be only 2 calls to `upper_bound`. This change won't complicate anything, it will only simplify. Assuming we want this change, maybe the current code is acceptable?

Reply:
> Assuming we want this change, maybe the current code is acceptable?

After more consideration I think my previous suggestion was wrong, I don't think we should do this change.

In truth, there is no good reason to disallow writes far into the past, as far as the user wants, if the coordinator knows what generation the write belongs to.

**If** someone is reading the CDC log table using a moving time window, then it is **on them** to synchronize their writes' timestamps with the moving time window so that the time window does not skip any such writes. It is not the responsibility of Scylla to protect the user from themselves in that case. And disallowing writes to the past eliminates valid workloads that do not depend on such moving time window, which could for example involve reading the full CDC log from time to time.

I think the only valid reason that we want to disallow writes into the past *if there was a generation switch in-between past and now*, is so that we keep as few generations in-memory as possible. Generations are pretty large data structures and it's preferable to keep at most one in memory, times where we keep more than one should be short. Hence the garbage-collection mechanism. It's not that we disallow writes into the past; it's that we garbage-collect past generations, hence we don't know where this write belongs.


---

Regarding the implementation and `upper_bound`s.

The previous version of the code started with `gen_used_at(now)`, removed all gens up to it, and then moved forward to find `gen_used_at(ts)` using this loop:
```cpp
    {
        auto next_it = std::next(it);
        while (next_it != _gens.end() && next_it->first <= ts) {
            it = next_it++;
        }
    }
```

so... wouldn't simply replacing `gen_used_at(now)`  with `gen_used_at(now - generation_leeway)`, and keeping the rest of the logic the same, achieve the result we want?

Reply:
And lemme highlight again that this
>  We can do it without losing data consistency because clients read CDC log tables with at least 10 s delay.

(from cover letter) is wrong. There is no data consistency issue at all. Again, if someone uses a workload with moving time window, where the time window may miss a write that was written behind the time window (and the window is only moving forward), it is not Scylla's fault. Scylla did save the write and the row did appear, there is no data inconsistency there.

Reply:
Data inconsistency from Scylla's side happens when Scylla picks the wrong generation. That can only happen when some coordinator doesn't know about a generation that the write belongs to, and incorrectly assumes that some old generation is still operating at that write's timestamp. That is why we need generation_leeway and choosing timestamps of generations in the future when creating new generations.

Reply:
> Regarding the implementation and `upper_bound`s.
> 
> The previous version of the code started with `gen_used_at(now)`, removed all gens up to it, and then moved forward to find `gen_used_at(ts)` using this loop:
> 
> ```c++
>     {
>         auto next_it = std::next(it);
>         while (next_it != _gens.end() && next_it->first <= ts) {
>             it = next_it++;
>         }
>     }
> ```
> 
> so... wouldn't simply replacing `gen_used_at(now)` with `gen_used_at(now - generation_leeway)`, and keeping the rest of the logic the same, achieve the result we want?

Again, there is a problem with the case `gen_used_at(ts) < gen_used_at(now - generation_leeway)`. We shouldn't search for `gen_used_at(ts)` in this case. We must handle it separately. Also, we don't know if `gen_used_at(ts)` exists (and is obsolete) or not. The current code throws different exceptions in both of these cases.

I think we should leave these two `gen_used_at` calls as they are. Generations at `ts` and `now - generation_leeway` are "independent". The first one can be older/equal/younger than the second one and they have independent roles in the code. The current code handles them independently. Code that calls `gen_used_at` once has to consider the order between these generations and gets more complicated and less readable.

We don't need the third `gen_used_at` call (the one from `auto current_gen = it == gen_used_at(now);`). It's sufficient to check if `gen_used_at(ts)` is one of the previous generations, which can be done by looking at the generation that follows it.

So, considering all of this, are two  `gen_used_at` calls acceptable?

Reply:
After some intense thinking, I managed to reduce the number of `gen_used_at` calls to 1 in a somewhat elegant way.

I still like the old code a bit more. I think it was easier to understand because it solved two problems -- garbage collection and accepting/rejecting a write -- independently.

The new code is "smarter" because it uses some properties, inequalities, etc. under the hood. Hence, it's less obvious why it (hopefully) works. I added some comments that clarify it. Anyway, since this code is on a hot path, I agree that it's a better choice.